packages/gatsby-source-contentful/src/types.js

```diff
@@ -0,0 +1,406 @@
+          },
+        },
+      }
+
+      return type
+    } else if (field.linkType === `Entry`) {
+      // Reference field
+      let contentTypes = null
+
+      // If there is validations field:
+      if (field.validations) {
+        // lok for validations that limit available content types
+        field.validations.some(validation => {
+          if (validation.linkContentType) {
+            contentTypes = context.contentTypeItems.filter(contentType =>
+              validation.linkContentType.includes(contentType.sys.contentful_id)
+            )
+            return true
+          }
+          return false
+        })
+      }
+
+      if (!contentTypes) {
+        contentTypes = context.contentTypeItems
+      }
+
+      // add backreferences
+      contentTypes.forEach(contentType => {
+        context.addBackReference({
+          referencingType: context.typeName,
+          referencedType: contentType.name,
+        })
+      })
+
+      if (contentTypes.length > 1) {
+        // this is union field
+        const usedTypesNames = contentTypes.map(contentType =>
+          makeTypeName(contentType.name)
+        )
+
+        // this is re-implementation of union type naming function from gatsby
+        const UnionTypeName = usedTypesNames.sort().join(``) + `Union`
+
+        let unionType = context.gqlTypes.find(
+          gqlType =>
+            gqlType.kind === `UNION` && gqlType.config.name === UnionTypeName
+        )
+
+        // create union type, if it doesn't exist yet
+        if (!unionType) {
+          unionType = context.schema.buildUnionType({
+            name: UnionTypeName,
+            types: usedTypesNames,
+            resolveType: node => node.internal.type,
+          })
+
+          context.gqlTypes.push(unionType)
+        }
+
+        // use that type
+        return {
+          type: UnionTypeName,
+          extensions: {
+            link: {
+              from: `${context.fieldName}___NODE`,
+            },
+          },
+        }
+      } else if (contentTypes.length === 1) {
+        // this is single type - not sure if this should be special case
+        // (using regular object type)
+        // or should create union with single type?
+        return {
+          type: makeTypeName(contentTypes[0].name),
+          extensions: {
+            link: {
+              from: `${context.fieldName}___NODE`,
+            },
+          },
+        }
+      } else {
+        return null
+      }
+    }
+    return null
+  },
+  // Number - integer
+  Integer: () => `Int`,
+  // Number - decimal
+  Number: () => `Float`,
+  Location: () => makeTypeName(`Location`),
+
+  // TO-DO figure out how to attach fields args / resolver properly.
+  // I'm importing private internals from gatsby which is no-no
+  Date: () => {
+    return {
+      type: `Date`,
+      extensions: {
+        dateformat: {},
+      },
+    }
+  },
+
+  RichText: (field, context) => {
+    const richTextTypeName = _.camelCase(
```

Comment:
If we only have json type here, why do we create a type at all? Is it for future use?

Reply:
Yes, to allow other transformers to kick in if needed

Reply:
Also https://github.com/gatsbyjs/gatsby/pull/12816#issuecomment-476250725 (and my next comment).

@Khaledgarbaya https://github.com/contentful/gatsby-transformer-contentful-richtext is deprecated and https://www.contentful.com/developers/docs/tutorials/general/rich-text-and-gatsby/ seems like proper/canonical way to work with Rich Text now? So I'm not sure about transformers here. Also my changes actually remove ability to use transformers as JSON is not being created as new node now (but I can change that, if transformers are still wanted)

Reply:
That's a valid point @pieh go for it